### PR TITLE
[fix] Email notification mail send exception handling #171

### LINKED ISSUE #171: [bug] Email notification to invalid user email breaks code execution
As reported by [Sparkxxx](https://github.com/Sparkxxx) on Gitter, if a user does not have a valid email address then email sent to that user will raise `SMTPRecipientsRefused` exception.

We need to handle this exception gracefully to make sure execution of code is not affected. 

Following is traceback shared by [Sparkxxx](https://github.com/Sparkxxx):

```
File "/usr/lib/python3.7/smtplib.py", line 881, in sendmail raise SMTPRecipientsRefused(senderrs) smtplib.SMTPRecipientsRefused: {'NOT@VALID.EMAIL': (550, b'5.1.1 NOT@VALID.EMAIL: Recipient address rejected: User unknown in local recipient table')}
```

Following code is responsible for sending email:
https://github.com/openwisp/openwisp-notifications/blob/d7f20ea41bcff1c0a437b93e98cac2d0a3bbec32/openwisp_notifications/handlers.py#L197-L213

### PR DESCRIPTION
**Pull request by RahulPalve**
Added exception handling in mail send, This will log any exception happend while sending mail

Closes #171

### CODE REVIEW

File: openwisp_notifications/handlers.py [hunk 1]
```diff
@@ -210,9 +210,14 @@ def send_email_notification(sender, instance, created, **kwargs):
...
-    mail.send()
-    # flag as emailed
-    instance.emailed = True
+    
+    try:
+        mail.send()
+        # flag as emailed
+        instance.emailed = True
+    except Exception as e:
+        logger.exception('Notification email sending failed :' + str(e))
```
Comment (Purhan) [inline openwisp_notifications/handlers.py:219]: We generally use the f-string convention:

```python
logger.exception(f'Notification email sending failed : {str(e)}')
```

**Review by Purhan — CHANGES_REQUESTED** [review]
Thanks @RahulPalve 
Please reference the issue properly in the commit message/PR description
Also check my comment below

**Review by pandafy — CHANGES_REQUESTED** [review]
@RahulPalve the build is failing due to QA issues. Locally, run `openwisp-qa-format` to format the code and `./run-qa-checks` to confirm your patch follows OpenWISP"s QA standards. You will have to fix those errors for CI build to pass. 

Read ["How to commit your changes properly" section of OpenWISP's contribution guidelines](https://openwisp.io/docs/developer/contributing.html#how-to-commit-your-changes-properly) if you already haven't.

It will also be nice to have a test for such scenario. Let us know if you need help writing test for this. If you get stuck, feel free to ask for help on [OpenWISP's Gitter channel](https://gitter.im/openwisp/general) or this PR.

**Review by pandafy — CHANGES_REQUESTED** [review]
Tests are missing. If you have any trouble in writing tests, feel free to ask for help in [OpenWISP's Gitter chat](https://gitter.im/openwisp/general).

File: openwisp_notifications/handlers.py [hunk 2]
```diff
@@ -215,10 +216,17 @@ def send_email_notification(sender, instance, created, **kwargs):
...
         )
         mail.attach_alternative(html_message, 'text/html')
-    mail.send()
-    # flag as emailed
-    instance.emailed = True
-    instance.save()
+
+    try:
+        mail.send()
+
```
Comment (pandafy) [inline openwisp_notifications/handlers.py:222]:
```suggestion
```

File: openwisp_notifications/handlers.py [hunk 3]
```diff
@@ -215,10 +216,17 @@ def send_email_notification(sender, instance, created, **kwargs):
...
+    except SMTPException as e:
+        logger.exception(f'Notification email sending failed : {str(e)}')
+
```
Comment (pandafy) [inline openwisp_notifications/handlers.py:225]:
```suggestion
```

**Review by pandafy — CHANGES_REQUESTED** [review]
Tests are missing. If you have any trouble in writing tests, feel free to ask for help in [OpenWISP's Gitter chat](https://gitter.im/openwisp/general).

**Comment by pandafy** [comment]
Closing PR due to lack of activity.

**Pull request closed without merging**